# About OOB Score

In [ ]:
1. OOB_Score is a very powerful Validation Technique used especially for the Random Forest algorithm for least Variance results.<br/>
2. Whenever we distribute data under the bootstrap with sampling process to base algorithum in Bagging. There are chances that some of the datapoints will never appear in any of the sub datasets provided to base ML models. The data points which are not included in any models are collected and an independent accuracy score on these data sets are conducted, that accuracy score is called **OOB Score**. Model is considered roboust if the model accuracy is near to OOB score. <br/>
3. In some way OOB Score helps in preventing **Data Interpretation leakage**.<br/>
4. However OOB Score is limited to Random Forest only.

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [2]:
df = pd.read_csv('/Users/manu/Downloads/100-days-of-machine-learning-main/day65-random-forest/heart.csv')

In [3]:
df.sample(3)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
137   62    1   1       128   208    1        0      140      0      0.0   
244   56    1   0       132   184    0        0      105      1      2.1   
211   61    1   0       120   260    0        1      140      1      3.6   

     slope  ca  thal  target  
137      2   0     2       1  
244      1   1     1       0  
211      1   1     3       0

In [4]:
X = df.iloc[:,0:-1]
y = df.iloc[:,-1]

In [5]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [6]:
rf = RandomForestClassifier(oob_score=True)

In [7]:
rf.fit(X_train,y_train)

RandomForestClassifier(oob_score=True)

In [8]:
rf.oob_score_

0.78099173553719

In [9]:
y_pred = rf.predict(X_test)
accuracy_score(y_test,y_pred)

0.8688524590163934

## Observation:

Our model should shell out an accuracy of 78% according to **OOB score**.<br/>
However the actual accuracy score observed is little more than OOB score which is good to have but we should be cautious as the accuracy is higher than 85%.<br/>
Sidenote : The accuracy for smaller data set is gernerally in between 80% - 85% and for large dataset it is in between 72%-75%.